### ISSUE TITLE
using one ScallopOption within another's definition breaks --help

### ISSUE THREAD

**Issue by hochgi** [scallop/scallop#143]
consider the following:
```scala
class Conf (arguments: Seq[String]) extends ScallopConf(arguments) {

  version("Example 1.0")
  val protocol = opt[String](name = "protocol", noshort = true, required = false, default = Some("http"))
  val host = opt[String](name = "host", short = 'h', required = true)
  val port = opt[Int](name = "port", noshort = true, default = Conf.defaultPortForProtocol(protocol()))
  verify()
}

object Conf {
  def defaultPortForProtocol(protocol: String): Option[Int] = protocol match {
    case "ftp" => Some(21)
    case "ssh" => Some(22)
    case "http" => Some(80)
    case "https" => Some(443)
    case _ => None
  }
}
```
this works as expected on regular runs, e.g:
```scala
object Main extends App {
  val conf = new Conf(args)
  val protocol = conf.protocol()
  val host = conf.host()
  val port = conf.port()
  
  println(s"protocol = $protocol")
  println(s"host = $host")
  println(s"port = $port")
}
```
and: `sbt "run --host google.com"`
will output:
```text
protocol = http
host = google.com
port = 80
```
but  `sbt "run --help"` yields an exception:
```text
[error] (run-main-0) org.rogach.scallop.exceptions.UnknownOption: Unknown option 'help'
org.rogach.scallop.exceptions.UnknownOption: Unknown option 'help'
	at org.rogach.scallop.Scallop.$anonfun$parse$16(Scallop.scala:158)
	at scala.Option.getOrElse(Option.scala:121)
	at org.rogach.scallop.Scallop.parse(Scallop.scala:158)
	at org.rogach.scallop.Scallop.parse(Scallop.scala:88)
	at org.rogach.scallop.Scallop.parsed$lzycompute(Scallop.scala:219)
	at org.rogach.scallop.Scallop.parsed(Scallop.scala:219)
	at org.rogach.scallop.Scallop.$anonfun$get$7(Scallop.scala:528)
	at scala.Option.map(Option.scala:146)
	at org.rogach.scallop.Scallop.get(Scallop.scala:527)
	at org.rogach.scallop.ScallopConfBase$$anon$1.$anonfun$fn$1(ScallopConfBase.scala:151)
	at org.rogach.scallop.ScallopOption.value$lzycompute(ScallopOption.scala:20)
	at org.rogach.scallop.ScallopOption.value(ScallopOption.scala:20)
	at org.rogach.scallop.ScallopOption.apply(ScallopOption.scala:30)
	at Conf.$anonfun$port$1(Conf.scala:14)
	at org.rogach.scallop.SimpleOption.helpInfo(CliOptions.scala:71)
	at org.rogach.scallop.ScallopHelpFormatter.$anonfun$getOptionsHelp$1(ScallopHelpFormatter.scala:21)
	at scala.collection.immutable.List.flatMap(List.scala:335)
	at org.rogach.scallop.ScallopHelpFormatter.getOptionsHelp(ScallopHelpFormatter.scala:19)
	at org.rogach.scallop.ScallopHelpFormatter.formatHelp(ScallopHelpFormatter.scala:7)
	at org.rogach.scallop.Scallop.help(Scallop.scala:473)
	at org.rogach.scallop.Scallop.printHelp(Scallop.scala:479)
	at org.rogach.scallop.ScallopConfBase.onError(ScallopConfBase.scala:400)
	at org.rogach.scallop.ScallopConfBase.verifyBuilder(ScallopConfBase.scala:361)
	at org.rogach.scallop.ScallopConfBase.verify(ScallopConfBase.scala:653)
	at Conf.<init>(Conf.scala:19)
	at Main$.delayedEndpoint$Main$1(Main.scala:12)
	at Main$delayedInit$body.apply(Main.scala:10)
	at scala.Function0.apply$mcV$sp(Function0.scala:34)
	at scala.Function0.apply$mcV$sp$(Function0.scala:34)
	at scala.runtime.AbstractFunction0.apply$mcV$sp(AbstractFunction0.scala:12)
	at scala.App.$anonfun$main$1$adapted(App.scala:76)
	at scala.collection.immutable.List.foreach(List.scala:389)
	at scala.App.main(App.scala:76)
	at scala.App.main$(App.scala:74)
	at Main$.main(Main.scala:10)
	at Main.main(Main.scala)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
```

tested with scala 2.12.2, scallop 3.0.1, sbt 0.13.15

**Comment by Rogach** [comment]
The problem here is that help is appending `(default = ...)` to option description line if `appendDefaultToDescription` is set. And it is not lazy, so default is computed even if `appendDefaultToDescription` is false (that's the default setting).

I'll fix the laziness problem in the next version.

You can also use this workaround:

```scala
class Conf (arguments: Seq[String]) extends ScallopConf(arguments) {
  val protocol = opt[String](default = Some("http"))
  val port = opt[Int]().orElse(protocol.toOption.flatMap {
    case "http" => Some(80)
    case _ => None
  })
  verify()
}

val conf = new Conf(Seq("--help"))
conf.port shouldEqual 80
```
(but it won't print the default even if `appendDefaultToDescription` is set, and I don't think there is a way around it, as long as we allow arbitrary functions in `default` resolver)

**Comment by hochgi** [comment]
Wow. you're fast. Thanks!
regarding `appendDefaultToDescription`, I don't see a way you can automatically print a useful default.
When a user provides `--help`, he won't provide e.g `protocol`, and you can't know what to print.
Maybe the best approach would be to delegate responsibility to the user. 
like changing default to accept either an optional value, or a string for help provided by the user.

**Comment by Rogach** [comment]
Well, string for help can be already provided in the description (actually, that's the original reason for deprecating automatic `defauilt = ...` appending and creating optional `appendDefaultToDescription` setting).

Released `3.0.2`, should hit Maven Central within an hour.

**Comment by hochgi** [comment]
Thank you very much!

**Comment by hochgi** [comment]
@Rogach I updated to the new release.
your workaround works of course, but original code fails with:
```text
[scallop] Error: It seems you tried to get option value before you constructed all options (maybe you forgot to call .verify method?). Please, move all extraction of values to after 'verify' method in ScallopConf.
```
This happens when I don't provide `--port` regardless if `--protocol` was provided.
Is this is the intended behavior?
because in 3.0.1 it worked and only help threw exception.
(help works fine now)

**Comment by Rogach** [comment]
That's a bug. I tried to catch cases when options were accessed before verification was finished, but now it also catches valid cases.

**Comment by Rogach** [comment]
Fixed it - released `3.0.3`.

**Comment by hochgi** [comment]
thank you! again! 
verified everything works as expected.
